File: mobile-customer-app/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as io;
import '../core/constants/api_constants.dart';

// Customer-side counterpart to the driver app's SocketService - listens
// instead of broadcasting. Same event contract as
// backend/src/sockets/tracking.socket.js.
class SocketService {
  io.Socket? _socket;

  bool get isConnected => _socket?.connected ?? false;

  void connect(String accessToken) {
    if (_socket != null) return;
    _socket = io.io(
      ApiConstants.socketUrl,
      io.OptionBuilder()
          .setTransports(['websocket'])
          .setAuth({'token': accessToken})
          .disableAutoConnect()
          .build(),
    );
    _socket!.connect();
  }

  void joinBookingRoom(String bookingId) {
    _socket?.emit('join_booking_room', {'bookingId': bookingId});
  }

  void leaveBookingRoom(String bookingId) {
    _socket?.emit('leave_booking_room', {'bookingId': bookingId});
  }

  void onLocationBroadcast(void Function(Map<String, dynamic>) callback) {
    _socket?.on('location_broadcast', (data) => callback(Map<String, dynamic>.from(data as Map)));
  }

  void onStatusBroadcast(void Function(Map<String, dynamic>) callback) {
    _socket?.on('status_broadcast', (data) => callback(Map<String, dynamic>.from(data as Map)));
  }

  void sendCustomerLocation(String bookingId, double lat, double lng) {
    _socket?.emit('customer_location_update', {'bookingId': bookingId, 'lat': lat, 'lng': lng});
  }

  void sendChatMessage(String bookingId, String text) {
    _socket?.emit('chat_message', {'bookingId': bookingId, 'text': text});
  }

  void onChatMessage(void Function(Map<String, dynamic>) callback) {
    _socket?.on('chat_message', (data) => callback(Map<String, dynamic>.from(data as Map)));
  }

  void dispose() {
    _socket?.disconnect();
    _socket?.dispose();
    _socket = null;
  }
}

File: web-portal-admin/src/shared/services/socketService.js
import { io } from 'socket.io-client';
import { useAuthStore } from '../store/authStore.js';

// Mirrors the (fixed-this-session) mobile-driver-app socket_service.dart
// pattern: a singleton connection that guards on `.connected`, not just
// non-null, so a stale/disconnected socket from a previous session never
// silently blocks reconnecting with a fresh token. Admin only ever listens
// (view order tracking/chat) - it never emits driver_location_update,
// customer_location_update, or chat_message, since the backend's
// tracking.socket.js role-gates all three to 'driver'/'customer' senders.
function socketOrigin() {
  const apiBase = import.meta.env.VITE_API_BASE_URL || 'http://localhost:5000/api/v1';
  return apiBase.replace(/\/api\/v1\/?$/, '');
}

let socket = null;

function connect() {
  if (socket && socket.connected) return socket;
  socket?.disconnect();
  const { accessToken } = useAuthStore.getState();
  socket = io(socketOrigin(), {
    transports: ['websocket'],
    auth: { token: accessToken },
    autoConnect: false,
  });
  socket.connect();
  return socket;
}

function joinBookingRoom(bookingId) {
  socket?.emit('join_booking_room', { bookingId });
}

function leaveBookingRoom(bookingId) {
  socket?.emit('leave_booking_room', { bookingId });
}

function onLocationBroadcast(callback) {
  socket?.on('location_broadcast', callback);
}

function onChatMessage(callback) {
  socket?.on('chat_message', callback);
}

function onConnectionChange(callback) {
  socket?.on('connect', () => callback(true));
  socket?.on('disconnect', () => callback(false));
  socket?.on('connect_error', () => callback(false));
}

function dispose() {
  socket?.disconnect();
  socket = null;
}

export const socketService = {
  connect,
  joinBookingRoom,
  leaveBookingRoom,
  onLocationBroadcast,
  onChatMessage,
  onConnectionChange,
  dispose,
};
